### PR TITLE
REF: Deduplicate to_xml code

### PR DESCRIPTION
**Pull request by phofl**
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them

This also fixes some typing bugs detected after removing the _init_ implementation

### CODE REVIEW

File: pandas/io/formats/xml.py [hunk 1]
```diff
@@ -247,17 +251,37 @@ def other_namespaces(self) -> dict:
 
         return nmsp_dict
 
-    def build_attribs(self) -> None:
+    def build_attribs(self, d: dict[str, Any], elem_row: Any) -> None:
```
Comment (jreback) [inline pandas/io/formats/xml.py:254]: don't luv the modification inline here, can you copy and return a new object?

**Review by jreback — CHANGES_REQUESTED** [review]
@ParfaitG if you can have a look

File: pandas/io/formats/xml.py (hunk 1 above)
Comment (phofl) [reply pandas/io/formats/xml.py]: We could return it like that

Comment (jreback) [reply pandas/io/formats/xml.py]: great, yeah i think that's a nicer pattern

Comment (phofl) [reply pandas/io/formats/xml.py]: Yep definitively

**Review by ParfaitG — COMMENTED** [review]
Thanks @phofl with refactoring this code! I just have a few comments in my walk-through.

File: pandas/io/formats/xml.py [hunk 2]
```diff
@@ -108,7 +107,7 @@ def __init__(
         encoding: str = "utf-8",
         xml_declaration: bool | None = True,
         pretty_print: bool | None = True,
-        stylesheet: FilePath | ReadBuffer[str] | None = None,
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/io/formats/xml.py:111]: No not sure. I assumed the types for the public API were correct, hence I adjusted this accordingly.

Turned that back to Read

File: pandas/tests/io/xml/test_to_xml.py [hunk 3]
```diff
@@ -1309,7 +1309,7 @@ def test_filename_and_suffix_comp(parser, compression_only):
 
 
 @td.skip_if_no("lxml")
-def test_ea_dtypes(any_numeric_ea_dtype):
+def test_ea_dtypes(any_numeric_ea_dtype, parser):
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/tests/io/xml/test_to_xml.py:1312]: Ah forgot to remove the skip. Yeah saw that too after starting the refactor

File: pandas/io/formats/xml.py [hunk 4]
```diff
@@ -251,7 +251,7 @@ def other_namespaces(self) -> dict:
 
         return nmsp_dict
 
-    def build_attribs(self, d: dict[str, Any], elem_row: Any) -> None:
+    def build_attribs(self, d: dict[str, Any], elem_row: Any) -> Any:
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/io/formats/xml.py:255]: Is not ideal, but they are tricky to type so I went wit Any for now

File: pandas/io/formats/xml.py [hunk 5]
```diff
@@ -247,17 +252,38 @@ def other_namespaces(self) -> dict:
...
-        raise AbstractMethodError(self)
+        if not self.attr_cols:
+            return elem_row
+
+        for col in self.attr_cols:
+            attr_name = self._get_flat_col_name(col)
+            try:
+                val = None if isna(d[col]) else str(d[col])
+                if val is not None:
+                    elem_row.attrib[attr_name] = val
```
Comment (jreback) [inline pandas/io/formats/xml.py:271]: should this be copied?

File: pandas/io/formats/xml.py [hunk 6]
```diff
@@ -311,16 +342,15 @@ def build_tree(self) -> bytes:
...
         for d in self.frame_dicts.values():
-            self.d = d
-            self.elem_row = SubElement(self.root, f"{self.prefix_uri}{self.row_name}")
+            elem_row = SubElement(self.root, f"{self.prefix_uri}{self.row_name}")
 
             if not self.attr_cols and not self.elem_cols:
-                self.elem_cols = list(self.d.keys())
-                self.build_elems()
+                self.elem_cols = list(d.keys())
+                self.build_elems(d, elem_row)
```
Comment (jreback) [inline pandas/io/formats/xml.py:349]: should this not assign?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/formats/xml.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ParfaitG) [reply pandas/io/formats/xml.py]: Maybe also add `ReadBuffer[bytes]` since XSLT is essentially XML?

File: pandas/io/formats/xml.py (hunk 6 above)
Comment (ParfaitG) [reply pandas/io/formats/xml.py:349]: `build_elems()` returns None. XML trees are built in place from root with elements added to it. After processing, the root (not any sub element) is passed to output.

File: pandas/io/formats/xml.py (hunk 5 above)
Comment (ParfaitG) [reply pandas/io/formats/xml.py:271]: Maybe avoid `val`? (This change came on last recent `to_xml` PR).

```python
if not isna(d[col]):
    elem_row.attrib[attr_name] = str(d[col])
```

Comment (phofl) [reply pandas/io/formats/xml.py:271]: Yep thx, this sounds good

File: pandas/io/formats/xml.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ParfaitG) [reply pandas/io/formats/xml.py]: @phofl, did you see above suggestion?

[the remark this answers is not in the document]
Comment (phofl) [reply pandas/io/formats/xml.py:111]: Thx, missed it

**Comment by ParfaitG** [comment]
LGTM! Thanks, @phofl. Be sure to rebase.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl and @ParfaitG for reviewing

**Pull request merged**